File: Baekjoon/분할정복/2630번_색종이_만들기.py
import sys

global blue
global white
blue = 0
white = 0

def cut(p, r_index, c_index, length):
  global blue
  global white

  if length == 1:
    if p[r_index][c_index] == 0:
      blue += 1
    else:
      white += 1
      return
  
  color = p[r_index][c_index]

  for i in range(r_index, r_index + length):
    for j in range(c_index, c_index + length):
      if p[i][j] != color:
        cut(p, r_index, c_index, length // 2)
        cut(p, r_index + length // 2, c_index, length // 2)
        cut(p, r_index, c_index + length // 2, length // 2)
        cut(p, r_index + length // 2, c_index + length // 2, length // 2)
  
  if p[r_index][c_index] == 0:
    blue += 1
  else:
    white += 1
    return

paper = []
n = int(input())

for i in range(n):
  paper.append(sys.stdin.readline()[:-1])

print(white)
print(blue)

File: Baekjoon/분할정복/2630번_색종이_만들기.c
#include <stdio.h>

int blue = 0, white = 0;

void cut(int **p, int r, int c, int len) {\
  if (len == 1) {
    if (p[r][c] == 0)
      blue++;
    else
      white++;
    return;
  }

  int color = p[r][c];

  for (int i = r; i < r + len; i++) {
    for (int j = c; j < c + len; j++) {
      if (p[i][j] != color) {
        cut(p, r, c, len / 2);  cut(p, r + len / 2, c, len / 2);
        cut(p, r, c + len / 2, len / 2);  cut(p, r + len / 2, c + len / 2, len / 2);
        return;
      }
    }
  }

  if (p[r][c] == 0)
    blue++;
  else
    white++;
  return;
}

int main() {
  int n;  scanf("%d", &n);
  int **paper = (int**)malloc(sizeof(int*)*n);
  for (int i = 0; i < n; i++)
    paper[i] = (int*)malloc(sizeof(int)*n);

  for (int i = 0; i < n; i++)
    for (int j = 0; j < n; j++)
      scanf("%d", &paper[i][j]);
  
  cut(paper, 0, 0, n);

  printf("%d\n%d\n", blue, white);
  return 0;
}
